[iPadOS] Google Docs font picker is sometimes dismissed upon first click with trackpad

whsieh opened the pull request:
#### 0e244717e96a87a202301283de267755faafb192
<pre>
[iPadOS] Google Docs font picker is sometimes dismissed upon first click with trackpad
<a href="https://bugs.webkit.org/show_bug.cgi?id=242520">https://bugs.webkit.org/show_bug.cgi?id=242520</a>
rdar://94527350

Reviewed by Devin Rousso.

When clicking with a trackpad on iPad, we currently dispatch mouse events in the gesture
recognizer&apos;s target action (in this case, `-[WKContentView mouseGestureRecognizerChanged:]`).
However, this action selector is invoked when UIKit routes both `UITouchesEvent` and `UIHoverEvent`
through the application (and the window containing the web view). These events are propagated from
backboard in the form of HID events, and may come in either order when clicking; in the case where
the `UIHoverEvent` precedes the `UITouchesEvent`, we end up with a &quot;mousemove&quot; followed by
&quot;mousedown&quot;. However, in the case where the `UITouchesEvent` precedes `UIHoverEvent`, the last mouse
event (of `MouseDown` type) gets created underneath `-touchesBegan:withEvent:` and is dispatched
twice.

This duplicate &quot;mousedown&quot; event results in Google Docs toggling some of its clickable menus twice
when clicking, which results in these menus opening and immediately closing (or vice versa). To
avoid this, we rename `-lastMouseEvent` to `-takeLastMouseEvent` and adjust it so that it clears out
the `_lastEvent`.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView mouseGestureRecognizerChanged:]):
* Source/WebKit/UIProcess/ios/WKMouseGestureRecognizer.h:
* Source/WebKit/UIProcess/ios/WKMouseGestureRecognizer.mm:
(-[WKMouseGestureRecognizer takeLastMouseEvent]):
(-[WKMouseGestureRecognizer lastMouseEvent]): Deleted.

Canonical link: <a href="https://commits.webkit.org/252290@main">https://commits.webkit.org/252290@main</a>
</pre>

dcrousso commented on Source/WebKit/UIProcess/ios/WKMouseGestureRecognizer.mm:
```diff
@@ -90,8 +92,12 @@ - (BOOL)_shouldReceiveTouch:(UITouch *)touch forEvent:(UIEvent *)event recognize
+    if (!_lastEvent || (_lastEvent->type() == WebKit::WebEvent::MouseDown && _lastDispatchedEventIsMouseDown))
+        return nullptr;
+
+    _lastDispatchedEventIsMouseDown = _lastEvent->type() == WebKit::WebEvent::MouseDown;
```
should we also compare the `timestamp` and/or `position` just to make sure we're not somehow ruining a scenario where there are two `"mousedown"` separated by time/position in a row?

whsieh replied:
What scenario do you have in mind?

dcrousso replied:
i dunno maybe some sort of "mousedown, switch apps, mouseup, switch back, mousedown"?

i'm not saying that this exists today, but i'd hate for this logic to cause us issues in the future if it does suddenly become possible

whsieh replied:
I see. I'll double check that scenario, but I would expect to see a cancel event which should make this work. I also don't think checking timestamp or location is quite the right way to solve this problem, either, since both are somewhat arbitrary.

One thing I can do here is go for the other approach I described in the commit message, where we just take the last mouse event. (This results in a cleaner implementation as well, but I opted to not go for this because I deemed it slightly riskier).

dcrousso replied:
FWIW i feel like from a code readability perspective something like `takeLastMouseEvent` would be much cleaner (and would assuage my concerns about `prepareToDispatchMouseEvent`)

dcrousso commented on Source/WebKit/UIProcess/ios/WKMouseGestureRecognizer.mm:
```diff
@@ -90,8 +92,12 @@ - (BOOL)_shouldReceiveTouch:(UITouch *)touch forEvent:(UIEvent *)event recognize
 }
 
-- (WebKit::NativeWebMouseEvent *)lastMouseEvent
+- (WebKit::NativeWebMouseEvent *)prepareToDispatchMouseEvent
```
something about this name feels weird to me.  i wouldn't expect it to have a return value based on the name (i.e. it feels more like a "hey object please do this thing" rather than a "hey object give me a value").

also, this new logic also means that it can't be called twice in a row (such as if someone doesn't wanna save the return value to a variable the first time (e.g. how `WebPageProxy::currentlyProcessedMouseDownEvent` is used).  i feel like we should move this logic to some other method (e.g. `shouldHandleEvent`) so that we don't prevent `_lastEvent` from being accessed

whsieh replied:
The intent is that this should only be called when we're about to dispatch a mouse event, which I _think_ should be clear from the name. I think that if someone (outside of `WKMouseGestureRecognizer`) wanted to just poll the last mouse event, we'd re-expose `-lastMouseEvent` as a readonly property instead. Currently, nothing requires this.

whsieh committed:
[iPadOS] Google Docs font picker is sometimes dismissed upon first click with trackpad
https://bugs.webkit.org/show_bug.cgi?id=242520
rdar://94527350

Reviewed by Devin Rousso.

When clicking with a trackpad on iPad, we currently dispatch mouse events in the gesture
recognizer's target action (in this case, `-[WKContentView mouseGestureRecognizerChanged:]`).
However, this action selector is invoked when UIKit routes both `UITouchesEvent` and `UIHoverEvent`
through the application (and the window containing the web view). These events are propagated from
backboard in the form of HID events, and may come in either order when clicking; in the case where
the `UIHoverEvent` precedes the `UITouchesEvent`, we end up with a "mousemove" followed by
"mousedown". However, in the case where the `UITouchesEvent` precedes `UIHoverEvent`, the last mouse
event (of `MouseDown` type) gets created underneath `-touchesBegan:withEvent:` and is dispatched
twice.

This duplicate "mousedown" event results in Google Docs toggling some of its clickable menus twice
when clicking, which results in these menus opening and immediately closing (or vice versa). To
avoid this, we rename `-lastMouseEvent` to `-takeLastMouseEvent` and adjust it so that it clears out
the `_lastEvent`.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView mouseGestureRecognizerChanged:]):
* Source/WebKit/UIProcess/ios/WKMouseGestureRecognizer.h:
* Source/WebKit/UIProcess/ios/WKMouseGestureRecognizer.mm:
(-[WKMouseGestureRecognizer takeLastMouseEvent]):
(-[WKMouseGestureRecognizer lastMouseEvent]): Deleted.

Canonical link: https://commits.webkit.org/252290@main
